spark/v3.1/spark/src/main/scala/com/netease/arctic/spark/sql/execution/MergeRowsExec.scala

```diff
@@ -147,7 +146,7 @@ case class MergeRowsExec(
       }
     }
 
-    val processFunc: InternalRow => InternalRow = if (performCardinalityCheck) {
+    val processFunc: InternalRow => InternalRow = if (performCardinalityCheck || unMatchedRowCheck) {
       processRowWithCardinalityCheck
```

Comment:
consider changing the method name if you are doing check for both matched key and non-matched key.